qualtran/bloqs/qsp_bloq.py

```diff
@@ -0,0 +1,165 @@
+
+        $$
+        \begin{matrix}
+        e^{i(\lambda + \phi)} \cos(\theta) & e^{i\phi} \sin(\theta) \\
+        e^{i\phi} \sin(\theta) & - \cos(\theta)
+        \end{matrix}
+        $$
+
+    Returns:
+        A 2x2 rotation matrix
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Equation 7.
+    """
+    return np.array(
+        [
+            [np.exp(1j * (lambd + phi)) * np.cos(theta), np.exp(1j * phi) * np.sin(theta)],
+            [np.exp(1j * lambd) * np.sin(theta), -np.cos(theta)],
+        ]
+    )
+
+
+def qsp_phase_factors(
+    P: Sequence[complex], Q: Sequence[complex]
+) -> Tuple[Sequence[float], Sequence[float], float]:
+    """Computes the QSP signal rotations for a given pair of polynomials.
+
+    The QSP transformation is described in Theorem 3, and the algorithm for computing
+    co-efficients is described in Algorithm 1.
+
+    Args:
+        P: Co-efficients of a complex polynomial.
+        Q: Co-efficients of a complex polynomial.
+
+    Returns:
+        A tuple (theta, phi, lambda).
+        theta and phi have length degree(P) + 1.
+
+    Raises:
+        ValueError: when P and Q have different degrees.
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Theorem 3; Algorithm 1.
+    """
+    if len(P) != len(Q):
+        raise ValueError("Polynomials P and Q must have the same degree.")
+
+    S = np.array([P, Q])
+    n = S.shape[1]
+
+    theta = np.zeros(n)
+    phi = np.zeros(n)
+    lambd = 0
+
+    for d in range(n - 1, 0, -1):
+        assert S.shape == (2, d + 1)
+
+        a, b = S[:, d]
+        theta[d] = np.arctan2(np.abs(a), np.abs(b))
+        phi[d] = np.angle(a / b)
+
+        if d == 0:
+            lambd = np.angle(b)
+        else:
+            S = _arbitrary_SU2_rotation(theta[d], phi[d], 0) @ S
+            S = np.array([S[0][1:d], S[1][0 : d - 1]])
+
+    return theta, phi, lambd
+
+
+@frozen
+class GeneralizedQSP(GateWithRegisters):
+    r"""Applies a QSP polynomial $P$ to a unitary $U$ to obtain a block-encoding of $P(U)$.
+
+    Given a pair of QSP polynomials $P, Q$, this gate represents the following unitary:
+
+        $$ \begin{bmatrix} P(U) & \cdot \\ Q(U) & \cdot \end{bmatrix} $$
+
+    The polynomials $P, Q$ must satisfy:
+        $\abs{P(x)}^2 + \abs{Q(x)}^2 = 1$ for every $x \in \mathbb{C}$ such that $\abs{x} = 1$
+
+    The exact circuit is described in Figure 2.
+
+    Args:
+        U: Unitary operation.
+        P: Co-efficients of a complex polynomial.
+        Q: Co-efficients of a complex polynomial.
+
+    References:
+        [Generalized Quantum Signal Processing](https://arxiv.org/abs/2308.01501)
+            Motlagh and Wiebe. (2023). Theorem 3; Figure 2.
+    """
+
+    U: GateWithRegisters
+    P: Sequence[complex]
+    Q: Sequence[complex]
```

Comment:
can we use more meaningful and pep8 compliant variable names for these?

Reply:
Any suggestions? For `U`, Is there some naming convention for nested bloqs?

As for `P, Q`, I'm not really sure what to name them (as the literature also doesn't explicitly name them). Perhaps I could call them qsp_polynomials and store them as a tuple instead?